core/shared/src/main/scala/zio/ConfigProvider.scala

```diff
@@ -35,19 +35,26 @@ trait ConfigProvider { self =>
    * single configuration value.
    */
   final def nested(name: String): ConfigProvider =
-    new ConfigProvider {
-      def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A] = self.load(config.nested(name))
-    }
+    ConfigProvider.fromFlat(self.flatten.nested(name))
 
   /**
    * Returns a new config provider that preferentially loads configuration data
    * from this one, but which will fall back to the specified alterate provider
    * if there are any issues loading the configuration from this provider.
    */
   final def orElse(that: ConfigProvider): ConfigProvider =
-    new ConfigProvider {
-      def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A] =
-        self.load(config).orElse(that.load(config))
+    ConfigProvider.fromFlat(self.flatten.merge(that.flatten))
+
+  /**
+   * Flattens this config provider into a simplified config provider that knows
+   * only how to deal with flat (key/value) properties.
+   */
+  def flatten: ConfigProvider.Flat =
+    new ConfigProvider.Flat {
+      def load[A](path: Chunk[String], config: Config.Primitive[A])(implicit trace: Trace): IO[Config.Error, Chunk[A]] =
+        ZIO.die(new NotImplementedError("ConfigProvider#flatten"))
```

Comment:
Maybe it would be better to return `Option[ConfigProvider.Flat]` than `die`? I'm not sure.

Reply:
I think it depends on whether we think other `ConfigProvider` instances we define will be able to implement `flatten`. 

With the changes in this PR every `ConfigProvider` constructor and every `ConfigProvider` operator returns a `ConfigProvider` that implements `flatten`.

If we can continue this in ZIO Config then I think this failure really is an extraordinary situation and `die` provides a smoother interface where users of `flatten` don't have to handle the possibility of `None`.

On the other hand, if we think that we will have a significant number of `ConfigProvider` constructors that can't implement `flatten` then I think we would need to look at alternatives.